In [1]:
import pandas as pd
import numpy as np

In [2]:
df_house = pd.read_csv("data.csv")
df_house

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h        NaN      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0  ...       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0  ...       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0  ...       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0  ...       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0  ...       2.0  0.0   
...              ...         ...       ...       ...  ...       ...  ...   
23673      Harcourts  26-08-2017      27.2    3024.0  ...       1.0  0.0   
23674  hockingstuart  26-08-2017      27.2    3024.0  ...       NaN  NaN   
23675     Buckingham  26-08-2017      12.7    3085.0  ...       3.0  2.0   
23676        Village  26-08-2017       6.3    3013.0  ...       1.0  1.0   
23677        Village  26-08-2017       6.3    3013.0  ...       2.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0         126.0           NaN        NaN        Yarra -37.80140   144.99580   
1         202.0           NaN        NaN        Yarra -37.79960   144.99840   
2         156.0          79.0     1900.0        Yarra -37.80790   144.99340   
3           0.0           NaN        NaN        Yarra -37.81140   145.01160   
4         134.0         150.0     1900.0        Yarra -37.80930   144.99440   
...         ...           ...        ...          ...       ...         ...   
23673     552.0         119.0     1990.0          NaN -37.90032   144.61839   
23674       NaN           NaN        NaN          NaN -37.87882   144.60184   
23675       NaN           NaN        NaN          NaN -37.72006   145.10547   
23676     362.0         112.0     1920.0          NaN -37.81188   144.88449   
23677       NaN         139.0     1950.0          NaN -37.81829   144.87404   

                  Regionname Propertycount  
0      Northern Metropolitan        4019.0  
1      Northern Metropolitan        4019.0  
2      Northern Metropolitan        4019.0  
3      Northern Metropolitan        4019.0  
4      Northern Metropolitan        4019.0  
...                      ...           ...  
23673   Western Metropolitan        5262.0  
23674   Western Metropolitan        5262.0  
23675  Northern Metropolitan        1369.0  
23676   Western Metropolitan        6543.0  
23677   Western Metropolitan        6543.0  

[23678 rows x 21 columns]

In [3]:
df_house.shape

(23678, 21)

In [4]:
df_house.dtypes

Suburb            object
Address           object
Rooms            float64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

In [5]:
df_house.isnull().sum()

Suburb             131
Address            131
Rooms              131
Type               131
Price             5282
Method             131
SellerG            131
Date               131
Distance           132
Postcode           132
Bedroom2          4612
Bathroom          4615
Car               4757
Landsize          6268
BuildingArea     13660
YearBuilt        12138
CouncilArea       8022
Lattitude         4435
Longtitude        4435
Regionname         132
Propertycount      132
dtype: int64

Percentage of missing values column wise

In [6]:
df_house.isna().mean().round(4) * 100

Suburb            0.55
Address           0.55
Rooms             0.55
Type              0.55
Price            22.31
Method            0.55
SellerG           0.55
Date              0.55
Distance          0.56
Postcode          0.56
Bedroom2         19.48
Bathroom         19.49
Car              20.09
Landsize         26.47
BuildingArea     57.69
YearBuilt        51.26
CouncilArea      33.88
Lattitude        18.73
Longtitude       18.73
Regionname        0.56
Propertycount     0.56
dtype: float64

In [7]:
df_house

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h        NaN      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0  ...       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0  ...       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0  ...       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0  ...       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0  ...       2.0  0.0   
...              ...         ...       ...       ...  ...       ...  ...   
23673      Harcourts  26-08-2017      27.2    3024.0  ...       1.0  0.0   
23674  hockingstuart  26-08-2017      27.2    3024.0  ...       NaN  NaN   
23675     Buckingham  26-08-2017      12.7    3085.0  ...       3.0  2.0   
23676        Village  26-08-2017       6.3    3013.0  ...       1.0  1.0   
23677        Village  26-08-2017       6.3    3013.0  ...       2.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0         126.0           NaN        NaN        Yarra -37.80140   144.99580   
1         202.0           NaN        NaN        Yarra -37.79960   144.99840   
2         156.0          79.0     1900.0        Yarra -37.80790   144.99340   
3           0.0           NaN        NaN        Yarra -37.81140   145.01160   
4         134.0         150.0     1900.0        Yarra -37.80930   144.99440   
...         ...           ...        ...          ...       ...         ...   
23673     552.0         119.0     1990.0          NaN -37.90032   144.61839   
23674       NaN           NaN        NaN          NaN -37.87882   144.60184   
23675       NaN           NaN        NaN          NaN -37.72006   145.10547   
23676     362.0         112.0     1920.0          NaN -37.81188   144.88449   
23677       NaN         139.0     1950.0          NaN -37.81829   144.87404   

                  Regionname Propertycount  
0      Northern Metropolitan        4019.0  
1      Northern Metropolitan        4019.0  
2      Northern Metropolitan        4019.0  
3      Northern Metropolitan        4019.0  
4      Northern Metropolitan        4019.0  
...                      ...           ...  
23673   Western Metropolitan        5262.0  
23674   Western Metropolitan        5262.0  
23675  Northern Metropolitan        1369.0  
23676   Western Metropolitan        6543.0  
23677   Western Metropolitan        6543.0  

[23678 rows x 21 columns]

In [8]:
df_house.isnull().any(axis = 1).sum()

17482

In [9]:
len(df_house)

23678

In [10]:
del df_house['BuildingArea']

In [11]:
del df_house['YearBuilt']

In [12]:
del df_house['CouncilArea']

In [13]:
df_house

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h        NaN      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0       2.0       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0       3.0       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
...              ...         ...       ...       ...       ...       ...  ...   
23673      Harcourts  26-08-2017      27.2    3024.0       3.0       1.0  0.0   
23674  hockingstuart  26-08-2017      27.2    3024.0       NaN       NaN  NaN   
23675     Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23676        Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23677        Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0         126.0  -37.80140   144.99580  Northern Metropolitan         4019.0  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
3           0.0  -37.81140   145.01160  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23673     552.0  -37.90032   144.61839   Western Metropolitan         5262.0  
23674       NaN  -37.87882   144.60184   Western Metropolitan         5262.0  
23675       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23676     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23677       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[23678 rows x 18 columns]

In [14]:
df_house = df_house[df_house.isnull().sum(axis=1) < 5]

In [15]:
df_house

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23672  Williamstown        96 Verdon St    4.0    h  2500000.0     PI   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

          SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0          Jellis  03-09-2016       2.5    3067.0       2.0       1.0  1.0   
1          Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2          Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
3          Rounds  04-02-2016       2.5    3067.0       3.0       2.0  1.0   
4          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
...           ...         ...       ...       ...       ...       ...  ...   
23672     Sweeney  26-08-2017       6.8    3016.0       4.0       1.0  5.0   
23673   Harcourts  26-08-2017      27.2    3024.0       3.0       1.0  0.0   
23675  Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23676     Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23677     Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0         126.0  -37.80140   144.99580  Northern Metropolitan         4019.0  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
3           0.0  -37.81140   145.01160  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23672     866.0  -37.85908   144.89299   Western Metropolitan         6380.0  
23673     552.0  -37.90032   144.61839   Western Metropolitan         5262.0  
23675       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23676     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23677       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[19226 rows x 18 columns]

In [16]:
df_house.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price            4140
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2          160
Bathroom          163
Car               305
Landsize         1816
Lattitude          26
Longtitude         26
Regionname          0
Propertycount       0
dtype: int64

In [17]:
df_house

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23672  Williamstown        96 Verdon St    4.0    h  2500000.0     PI   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

          SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0          Jellis  03-09-2016       2.5    3067.0       2.0       1.0  1.0   
1          Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2          Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
3          Rounds  04-02-2016       2.5    3067.0       3.0       2.0  1.0   
4          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
...           ...         ...       ...       ...       ...       ...  ...   
23672     Sweeney  26-08-2017       6.8    3016.0       4.0       1.0  5.0   
23673   Harcourts  26-08-2017      27.2    3024.0       3.0       1.0  0.0   
23675  Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23676     Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23677     Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0         126.0  -37.80140   144.99580  Northern Metropolitan         4019.0  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
3           0.0  -37.81140   145.01160  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23672     866.0  -37.85908   144.89299   Western Metropolitan         6380.0  
23673     552.0  -37.90032   144.61839   Western Metropolitan         5262.0  
23675       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23676     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23677       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[19226 rows x 18 columns]

In [18]:
df_house['Bedroom2'].fillna((df_house['Price'].mean()), inplace=True)

/Users/vishaln/opt/anaconda3/lib/python3.7/site-packages/pandas/core/generic.py:6287: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [19]:
df_house

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23672  Williamstown        96 Verdon St    4.0    h  2500000.0     PI   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

          SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0          Jellis  03-09-2016       2.5    3067.0       2.0       1.0  1.0   
1          Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2          Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
3          Rounds  04-02-2016       2.5    3067.0       3.0       2.0  1.0   
4          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
...           ...         ...       ...       ...       ...       ...  ...   
23672     Sweeney  26-08-2017       6.8    3016.0       4.0       1.0  5.0   
23673   Harcourts  26-08-2017      27.2    3024.0       3.0       1.0  0.0   
23675  Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23676     Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23677     Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0         126.0  -37.80140   144.99580  Northern Metropolitan         4019.0  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
3           0.0  -37.81140   145.01160  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23672     866.0  -37.85908   144.89299   Western Metropolitan         6380.0  
23673     552.0  -37.90032   144.61839   Western Metropolitan         5262.0  
23675       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23676     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23677       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[19226 rows x 18 columns]

In [20]:
df_house.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price            4140
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom          163
Car               305
Landsize         1816
Lattitude          26
Longtitude         26
Regionname          0
Propertycount       0
dtype: int64

In [21]:
df_house['Car'].fillna((df_house['Price'].mean()), inplace=True)

In [22]:
df_house.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price            4140
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom          163
Car                 0
Landsize         1816
Lattitude          26
Longtitude         26
Regionname          0
Propertycount       0
dtype: int64

In [23]:
df_house.shape

(19226, 18)

Percentage of missing values still left column wise

In [24]:
df_house.isna().mean().round(4) * 100

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.53
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          0.00
Bathroom          0.85
Car               0.00
Landsize          9.45
Lattitude         0.14
Longtitude        0.14
Regionname        0.00
Propertycount     0.00
dtype: float64

In [25]:
df_house['Price'].fillna((df_house['Price'].mean()), inplace=True)

Explanation to treat missing values in target variable -  I have imputed the NaN in target variable with mean value so that 
the NaN values are treated in the appropriate manner. 

In [26]:
df_house.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom          163
Car                 0
Landsize         1816
Lattitude          26
Longtitude         26
Regionname          0
Propertycount       0
dtype: int64

In [27]:
df_house['Bathroom'].fillna((df_house['Price'].mean()), inplace=True)

In [28]:
df_house['Landsize'].fillna((df_house['Price'].mean()), inplace=True)

In [29]:
df_house['Lattitude'].fillna((df_house['Price'].mean()), inplace=True)

In [30]:
df_house['Longtitude'].fillna((df_house['Price'].mean()), inplace=True)

In [31]:
df_house.isnull().sum()

Suburb           0
Address          0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
Lattitude        0
Longtitude       0
Regionname       0
Propertycount    0
dtype: int64

In [32]:
df_house.shape

(19226, 18)

Percentage of rows removed = 23678 - 19226 = 4452 / 23678 = 18.8%

Percentage of columns removed = 21 - 18 = 3 / 21 = 14.2%

Justification for row removal - I have removed almost 18.8% rows in order to completely clear all the missing values in the date.
There are so many missing values in the dataset which essentially need to be cleaned. 